src/job-cache/job-cache.cpp

```diff
@@ -90,25 +138,44 @@ static void copy(int src_fd, int dst_fd) {
   }
 }
 
+// For older linux distros use Linux's sendfile
 #else
+
+#include <sys/sendfile.h>
+// This function just uses `sendfile` to make
+// an efficent copy. It is however not atomic because
+// we have to `fstat` before we call `sendfile`.
+// This means that if an external party decides to mutate
+// the file (espeically changing its size) then this function
+// will not work as intended. External parties *are* allowed to
+// unlink this file but they're not allowed to modify it in any
+// other way or else this function will race with that external
+// modification.
+
 static void copy(int src_fd, int dst_fd) {
-  // TODO: Write a fallback. This should only be needed for centos7.6
-  assert(false);
+  struct stat buf = {};
+  // There's a race here between the fstat and the copy_file_range
+  if (fstat(src_fd, &buf) < 0) {
+    log_fatal("fstat(src_fd = %d): %s", src_fd, strerror(errno));
+  }
+  off_t idx = 0;
+  size_t size = buf.st_size;
+  do {
+    intptr_t written = sendfile(dst_fd, src_fd, &idx, size);
+    if (written < 0) {
+      log_fatal("sendfile(src_fd = %d, NULL, dst_fd = %d, size = %d, 0): %s", src_fd, dst_fd,
+                buf.st_size, strerror(errno));
+    }
+    idx = written;
+    size -= written;
+  } while (size != 0);
```

Comment:
this looks eerily similar to writing a raw NAND device driver lol

Reply:
heh. Hopefully we don't have the issue where the first 4 physical blocks were being used for critical stuff and they were getting overloaded because the designers were like "it will probably never have that many cycles on it" and then all of our infra started to fail one by one at the exact same time...yeah that wasn't fun.

Reply:
👀